### ISSUE TITLE
v0.14.0 generates invalid negotiate tokens

### ISSUE THREAD

**Issue by twAlexM** [requests/requests-kerberos#173]
Hi team,

I'm seeing what appears to be a regression between requests-kerberos v0.11.0 and v0.14.0 that prevents successful authentication. To wit, a simple request to our internal test server with v0.11.0 succeeds:

```
Python 3.7.9 (default, Sep 23 2020, 09:52:36) 
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import logging
>>> logging.getLogger("requests_kerberos").setLevel(logging.DEBUG)
>>> logging.getLogger("requests_kerberos.kerberos_").setLevel(logging.DEBUG)
>>> logging.basicConfig(level=logging.DEBUG,format='%(relativeCreated)6d %(message)s')
>>> import requests_kerberos
>>> requests_kerberos.__version__
'0.11.0'
>>> import requests
>>> r = requests.get("https://krb-test.twitter.biz/", auth=requests_kerberos.HTTPKerberosAuth())
 20024 Starting new HTTPS connection (1): krb-test.twitter.biz:443
 20139 https://krb-test.twitter.biz:443 "GET / HTTP/1.1" 401 503
 20140 handle_401(): Handling: 401
 20161 authenticate_user(): Authorization header: Negotiate YIIDNgYJKoZIhvcSAQICAQBuggMlMIIDIaADAgEFoQMCAQ6iBwMFAKAAAACjggFcYYIBWDCCAVSgAwIBBaENGwtUV0lUVEVSLkJJWqInMCWgAwIBA6EeMBwbBEhUVFAbFGtyYi10ZXN0LnR3aXR0ZXIuYml6o4IBEzCCAQ+gAwIBEqEDAgECooIBAQSB/uJF0UU2vnprFd6KsGjsrh6yGG41pEjvZe6aN5C+3vsE5tQAPyKUkzvOwIxagf478XjrfHmcOU2FkA2h//oemKjAMAjyWjVmSijYidVbEgcmyUhgUyImQuYL4QMUPWFis22TA+/Ja51bZSfUixFreqM+arbzGVX7Gvdknd8Ujenj8MF81C3esNZ4wkdxg7mmF8J0q+HPFXGy97AH/3wyEMTkEuhwsjvBChWIn5MB03Hm32G8x6RwnmgNFuFkaHBSoQnsOiuoI1PEVFgb+Sv09S12cPgOVh1m0GYrLT9dcln8vAQDvg2cpBe6F15bFk1WW7xfON236NxpMUY7F0kepIIBqjCCAaagAwIBEqKCAZ0EggGZVnu1ZEgoO10drNZzS1aifzYN1LcsY7MKgmznOV/1eVP2YFj78s2CpWcJ8Pr9IKokS7+bEr/u27CZ62UJ31NuldrQonHeyTBhfbICYExgDUfEdGOVhGFTF6fJO0ggvzi9eC8GHzbzp3+F4VbAI+mbR+PZLVoWTPP9smGKYFPnfZr58UQgV1AAsiuKDw/S8mRBYDVRQuI/3bOHSyIWty7g5YgZPkZAZ2zzlCMIWT7hTqmGg9V21zYCBzEymlg+TPWafkyBRA0QPkBWZmz8KRkVGnXG1KHlZnB4ITtzMAFNhVYd4gOQI3NWVyE3ktEzlOlsE1frxCtzj7paMFr4ctCW9sz2Plc1jEl6TGz6CHWFXdRWs876AVIIxY3Y7NdzNcqLMHx+KQ/YVo9moq33RybA/S06SvFf4VQZZ/wPKHFXEjsXks9iDZScxLgvwQWvjS6h20kiMya3lAvKE8BWJzr1TnBnysMwY5qy1PgcOx1sS+UWxXjUYwLOpUs41Z3tJ5ybp+920q8kEcICjbJtMr/BJuRg+NASu6y60Q==
 20161 Resetting dropped connection: krb-test.twitter.biz
 20277 https://krb-test.twitter.biz:443 "GET / HTTP/1.1" 200 299
 20278 authenticate_user(): returning <Response [200]>
 20278 handle_401(): returning <Response [200]>
 20278 handle_response(): returning <Response [200]>
 20278 handle_response() has seen 0 401 responses
 20278 handle_other(): Handling: 200
 20279 handle_other(): Authenticating the server
 20279 authenticate_server(): Authenticate header: YIGZBgkqhkiG9xIBAgICAG+BiTCBhqADAgEFoQMCAQ+iejB4oAMCARKicQRvo1JjF8fkXRNZWsX/2cayZlOvFU7lnnoZyfL9hc3PaqsnCd6rkigLoAtr+37xVQ2EZTZr32HVaPTt0UqZfkbSt5ILKeEpJTeAubhp5wX4PnifH8Qs+5fpn4PQfIG1BE5dJhU0EjNK14MLe/DpNfMg
 20279 authenticate_server(): returning <Response [200]>
 20279 handle_other(): returning <Response [200]>
 20279 handle_response(): returning <Response [200]>
```

The same request with 0.14.0 fails, implicating `pyspnego`:

```
Python 3.7.9 (default, Sep 23 2020, 09:52:36) 
[Clang 11.0.0 (clang-1100.0.33.17)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
(InteractiveConsole)
>>> import logging
>>> logging.getLogger("requests_kerberos").setLevel(logging.DEBUG)
>>> logging.getLogger("requests_kerberos.kerberos_").setLevel(logging.DEBUG)
>>> logging.basicConfig(level=logging.DEBUG,format='%(relativeCreated)6d %(message)s')
>>> import requests_kerberos
  4885 SSPI bindings not available, cannot use any SSPIProxy protocols: No module named 'spnego._sspi_raw.sspi'
>>> requests_kerberos.__version__
'0.14.0'
>>> import requests
>>> r = requests.get("https://krb-test.twitter.biz/", auth=requests_kerberos.HTTPKerberosAuth())
 13717 Starting new HTTPS connection (1): krb-test.twitter.biz:443
 13838 https://krb-test.twitter.biz:443 "GET / HTTP/1.1" 401 503
 13905 handle_401(): Handling: 401
 13959 GSSAPI does not support required the NTLM interfaces: Major (65536):  An unsupported mechanism was requested, Minor (0): unknown mech-code 0 for mech unknown
 13959 GSSAPI step input: 
 13966 GSSAPI step output: YIIDUQYJKoZIhvcSAQICAQBuggNAMIIDPKADAgEFoQMCAQ6iBwMFACAAAACjggFcYYIBWDCCAVSgAwIBBaENGwtUV0lUVEVSLkJJWqInMCWgAwIBA6EeMBwbBEhUVFAbFGtyYi10ZXN0LnR3aXR0ZXIuYml6o4IBEzCCAQ+gAwIBEqEDAgECooIBAQSB/uJF0UU2vnprFd6KsGjsrh6yGG41pEjvZe6aN5C+3vsE5tQAPyKUkzvOwIxagf478XjrfHmcOU2FkA2h//oemKjAMAjyWjVmSijYidVbEgcmyUhgUyImQuYL4QMUPWFis22TA+/Ja51bZSfUixFreqM+arbzGVX7Gvdknd8Ujenj8MF81C3esNZ4wkdxg7mmF8J0q+HPFXGy97AH/3wyEMTkEuhwsjvBChWIn5MB03Hm32G8x6RwnmgNFuFkaHBSoQnsOiuoI1PEVFgb+Sv09S12cPgOVh1m0GYrLT9dcln8vAQDvg2cpBe6F15bFk1WW7xfON236NxpMUY7F0kepIIBxTCCAcGgAwIBEqKCAbgEggG0g/GrXEM56byvoygbc7kzsBoXCZ98xyCZa+mz3LweUjfG7JS+zOPxp0cMr54mIwmUciKUBV+4aQiH21AQhFQ9bSWEXTxdnF9szQjnC2ItqlOi5Covw5EcERFCSWzTlq5LKdDTlJWpvC09yrEcF40KeuSr6yfTupY0E4Je/2e/Epi4fTIHDN8jxY1Ucq46pEPzDmmjdjY3cJBP+yk+t6KgCHYHq6aCVAceD8fHf9g2nF/DlaWTS1OcA4y4qndJccpDSb4AyXIxuhk76jDBr/YqsPj3k+2SuBx8JPUXGDpgePloAwJ6lbRcdz313wFLyO6Ru4ZxjppbjfOhXS03FWw8TquGaRH+cJZtGYuDfZf7lw4elIVeTmXzkBpYlaX09oglrcqlFKMAaN8U9PvegCQSpO8z9kUCK2Mel/uOZD3gTcFLDmE8ZNESdj/QgaV2KzYohdgl7c27Og7t+u19hMm96XXxq5XirWryeDF78nxFt6ZNamc6GXkAS8XnDkj5LPctd2QreknQPIfimU/Q3Sw/FxlwT7u9mNrKmHnPKRxFQf2hdkNIfzqm0da6oIMCHivPNSXnvg==
 13966 authenticate_user(): Authorization header: Negotiate YIIDUQYJKoZIhvcSAQICAQBuggNAMIIDPKADAgEFoQMCAQ6iBwMFACAAAACjggFcYYIBWDCCAVSgAwIBBaENGwtUV0lUVEVSLkJJWqInMCWgAwIBA6EeMBwbBEhUVFAbFGtyYi10ZXN0LnR3aXR0ZXIuYml6o4IBEzCCAQ+gAwIBEqEDAgECooIBAQSB/uJF0UU2vnprFd6KsGjsrh6yGG41pEjvZe6aN5C+3vsE5tQAPyKUkzvOwIxagf478XjrfHmcOU2FkA2h//oemKjAMAjyWjVmSijYidVbEgcmyUhgUyImQuYL4QMUPWFis22TA+/Ja51bZSfUixFreqM+arbzGVX7Gvdknd8Ujenj8MF81C3esNZ4wkdxg7mmF8J0q+HPFXGy97AH/3wyEMTkEuhwsjvBChWIn5MB03Hm32G8x6RwnmgNFuFkaHBSoQnsOiuoI1PEVFgb+Sv09S12cPgOVh1m0GYrLT9dcln8vAQDvg2cpBe6F15bFk1WW7xfON236NxpMUY7F0kepIIBxTCCAcGgAwIBEqKCAbgEggG0g/GrXEM56byvoygbc7kzsBoXCZ98xyCZa+mz3LweUjfG7JS+zOPxp0cMr54mIwmUciKUBV+4aQiH21AQhFQ9bSWEXTxdnF9szQjnC2ItqlOi5Covw5EcERFCSWzTlq5LKdDTlJWpvC09yrEcF40KeuSr6yfTupY0E4Je/2e/Epi4fTIHDN8jxY1Ucq46pEPzDmmjdjY3cJBP+yk+t6KgCHYHq6aCVAceD8fHf9g2nF/DlaWTS1OcA4y4qndJccpDSb4AyXIxuhk76jDBr/YqsPj3k+2SuBx8JPUXGDpgePloAwJ6lbRcdz313wFLyO6Ru4ZxjppbjfOhXS03FWw8TquGaRH+cJZtGYuDfZf7lw4elIVeTmXzkBpYlaX09oglrcqlFKMAaN8U9PvegCQSpO8z9kUCK2Mel/uOZD3gTcFLDmE8ZNESdj/QgaV2KzYohdgl7c27Og7t+u19hMm96XXxq5XirWryeDF78nxFt6ZNamc6GXkAS8XnDkj5LPctd2QreknQPIfimU/Q3Sw/FxlwT7u9mNrKmHnPKRxFQf2hdkNIfzqm0da6oIMCHivPNSXnvg==
 13967 Resetting dropped connection: krb-test.twitter.biz
 14093 https://krb-test.twitter.biz:443 "GET / HTTP/1.1" 401 503
 14094 authenticate_user(): returning <Response [401]>
 14094 handle_401(): returning <Response [401]>
 14094 handle_response(): returning <Response [401]>
 14094 handle_response() has seen 0 401 responses
 14094 handle_401(): Handling: 401
 14095 GSSAPI step input: YHIGCSqGSIb3EgECAgMAfmMwYaADAgEFoQMCAR6kERgPMjAyMjAxMTAyMjQyMDlapQUCAwDJ2qYDAgE8qQ0bC1RXSVRURVIuQklaqicwJaADAgEBoR4wHBsESFRUUBsUa3JiLXRlc3QudHdpdHRlci5iaXo=
 14097 generate_request_header(): ctx step failed:
Traceback (most recent call last):
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_context.py", line 55, in wrapper
    return func(*args, **kwargs)
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_gss.py", line 446, in step
    out_token = self._context.step(in_token)
  File "/Users/alexm/.pex/install/decorator-5.1.0-py3-none-any.whl.f4bfb5084a96473cc464661f2d5fa10516e9cd7c/decorator-5.1.0-py3-none-any.whl/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/_utils.py", line 155, in check_last_err
    return func(self, *args, **kwargs)
  File "/Users/alexm/.pex/install/decorator-5.1.0-py3-none-any.whl.f4bfb5084a96473cc464661f2d5fa10516e9cd7c/decorator-5.1.0-py3-none-any.whl/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/_utils.py", line 128, in catch_and_return_token
    return func(self, *args, **kwargs)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/sec_contexts.py", line 523, in step
    return self._initiator_step(token=token)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/sec_contexts.py", line 544, in _initiator_step
    token)
  File "gssapi/raw/sec_contexts.pyx", line 246, in gssapi.raw.sec_contexts.init_sec_context
gssapi.raw.exceptions.MalformedParameterError: Major (51183616): A parameter was malformed Miscellaneous failure (see text), Minor (0): Success

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/alexm/.pex/install/requests_kerberos-0.14.0-py2.py3-none-any.whl.e58b20f03b5f1a1ad213a09ba5526f3ddc01e47c/requests_kerberos-0.14.0-py2.py3-none-any.whl/requests_kerberos/kerberos_.py", line 227, in generate_request_header
    gss_response = ctx.step(in_token=negotiate_resp_value)
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_context.py", line 58, in wrapper
    raise SpnegoError(base_error=native_err, context_msg=context) from native_err
spnego.exceptions.SpnegoError: SpnegoError (4294967295): Major (51183616): A parameter was malformed Miscellaneous failure (see text), Minor (0): Success, Context: Processing security token
 14100 SpnegoError (4294967295): Major (51183616): A parameter was malformed Miscellaneous failure (see text), Minor (0): Success, Context: Processing security token
Traceback (most recent call last):
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_context.py", line 55, in wrapper
    return func(*args, **kwargs)
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_gss.py", line 446, in step
    out_token = self._context.step(in_token)
  File "/Users/alexm/.pex/install/decorator-5.1.0-py3-none-any.whl.f4bfb5084a96473cc464661f2d5fa10516e9cd7c/decorator-5.1.0-py3-none-any.whl/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/_utils.py", line 155, in check_last_err
    return func(self, *args, **kwargs)
  File "/Users/alexm/.pex/install/decorator-5.1.0-py3-none-any.whl.f4bfb5084a96473cc464661f2d5fa10516e9cd7c/decorator-5.1.0-py3-none-any.whl/decorator.py", line 232, in fun
    return caller(func, *(extras + args), **kw)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/_utils.py", line 128, in catch_and_return_token
    return func(self, *args, **kwargs)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/sec_contexts.py", line 523, in step
    return self._initiator_step(token=token)
  File "/Users/alexm/.pex/install/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl.4ca41f1669826b8a74cfa024d3a9441b4e2e43bb/gssapi-1.7.2-cp37-cp37m-macosx_10_9_x86_64.whl/gssapi/sec_contexts.py", line 544, in _initiator_step
    token)
  File "gssapi/raw/sec_contexts.pyx", line 246, in gssapi.raw.sec_contexts.init_sec_context
gssapi.raw.exceptions.MalformedParameterError: Major (51183616): A parameter was malformed Miscellaneous failure (see text), Minor (0): Success

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/alexm/.pex/install/requests_kerberos-0.14.0-py2.py3-none-any.whl.e58b20f03b5f1a1ad213a09ba5526f3ddc01e47c/requests_kerberos-0.14.0-py2.py3-none-any.whl/requests_kerberos/kerberos_.py", line 227, in generate_request_header
    gss_response = ctx.step(in_token=negotiate_resp_value)
  File "/Users/alexm/.pex/install/pyspnego-0.3.1-py2.py3-none-any.whl.56a2e639d0f8768a244bf01a919a359d7df8d5c7/pyspnego-0.3.1-py2.py3-none-any.whl/spnego/_context.py", line 58, in wrapper
    raise SpnegoError(base_error=native_err, context_msg=context) from native_err
spnego.exceptions.SpnegoError: SpnegoError (4294967295): Major (51183616): A parameter was malformed Miscellaneous failure (see text), Minor (0): Success, Context: Processing security token
 14101 handle_401(): returning <Response [401]>
 14101 handle_response(): returning <Response [401]>
 14101 handle_response() has seen 1 401 responses
 14101 handle_response(): returning 401 <Response [401]>
```

The 401 error on the client side comes with an error on the server:
```[Mon Jan 10 22:42:09.052037 2022] [auth_kerb:error] [pid 224941:tid 140553693796096] [client 203.0.113.42:57775] gss_accept_sec_context() failed: Unspecified GSS failure.  Minor code may provide more information (, Invalid field length in token)```

Currently we are working around this by using v0.11.0. Should we also open an issue with PySpnego?

**Comment by jborean93** [comment]
Hmm this is a bit puzzling, if you compare the original request that the older requests-kerberos send you get

```yaml
# python -m spnego --format yaml --token YIIDNgYJKoZ...

MessageType: SPNEGO InitialContextToken
Data:
  thisMech: Kerberos (1.2.840.113554.1.2.2)
  innerContextToken:
    MessageType: AP-REQ (14)
    Data:
      pvno: 5
      msg-type: AP-REQ (14)
      ap-options:
        raw: 160
        flags:
        - mutual-required (32)
        - reserved (128)
      ticket:
        tkt-vno: 5
        realm: DOMAIN.COM
        sname:
          name-type: NT-SRV-HST (3)
          name-string:
          - HTTP
          - krb-test.domain.com
        enc-part:
          etype: AES256_CTS_HMAC_SHA1_96 (18)
          kvno: 2
          cipher: E245D14536BE7A6B15DE8AB068ECAE1EB2186E35A448EF65EE9A3790BEDEFB04E6D4003F2294933BCEC08C5A81FE3BF178EB7C799C394D85900DA1FFFA1E98A8C03008F25A35664A28D889D55B120726C9486053222642E60BE103143D6162B36D9303EFC96B9D5B6527D48B116B7AA33E6AB6F31955FB1AF7649DDF148DE9E3F0C17CD42DDEB0D678C2477183B9A617C274ABE1CF1571B2F7B007FF7C3210C4E412E870B23BC10A15889F9301D371E6DF61BCC7A4709E680D16E164687052A109EC3A2BA82353C454581BF92BF4F52D7670F80E561D66D0662B2D3F5D7259FCBC0403BE0D9CA417BA175E5B164D565BBC5F38DDB7E8DC6931463B17491E
      authenticator:
        etype: AES256_CTS_HMAC_SHA1_96 (18)
        kvno:
        cipher: 567BB56448283B5D1DACD6734B56A27F360DD4B72C63B30A826CE7395FF57953F66058FBF2CD82A56709F0FAFD20AA244BBF9B12BFEEDBB099EB6509DF536E95DAD0A271DEC930617DB202604C600D47C474639584615317A7C93B4820BF38BD782F061F36F3A77F85E156C023E99B47E3D92D5A164CF3FDB2618A6053E77D9AF9F14420575000B22B8A0F0FD2F2644160355142E23FDDB3874B2216B72EE0E588193E4640676CF3942308593EE14EA98683D576D736020731329A583E4CF59A7E4C81440D103E4056666CFC2919151A75C6D4A1E5667078213B7330014D85561DE2039023735657213792D13394E96C1357EBC42B738FBA5A305AF872D096F6CCF63E57358C497A4C6CFA0875855DD456B3CEFA015208C58DD8ECD77335CA8B307C7E290FD8568F66A2ADF74726C0FD2D3A4AF15FE1541967FC0F287157123B1792CF620D949CC4B82FC105AF8D2EA1DB49223326B7940BCA13C056273AF54E7067CAC330639AB2D4F81C3B1D6C4BE516C578D46302CEA54B38D59DED279C9BA7EF76D2AF2411C2028DB26D32BFC126E460F8D012BBACBAD1
    RawData: ...
RawData: ...
```

Compared to what is sends today (with pyspnego)

```yaml
# python -m spnego --format yaml --token YIIDNgYJKoZ...

MessageType: SPNEGO InitialContextToken
Data:
  thisMech: Kerberos (1.2.840.113554.1.2.2)
  innerContextToken:
    MessageType: AP-REQ (14)
    Data:
      pvno: 5
      msg-type: AP-REQ (14)
      ap-options:
        raw: 32
        flags:
        - mutual-required (32)
      ticket:
        tkt-vno: 5
        realm: DOMAIN.COM
        sname:
          name-type: NT-SRV-HST (3)
          name-string:
          - HTTP
          - krb-test.domain.com
        enc-part:
          etype: AES256_CTS_HMAC_SHA1_96 (18)
          kvno: 2
          cipher: E245D14536BE7A6B15DE8AB068ECAE1EB2186E35A448EF65EE9A3790BEDEFB04E6D4003F2294933BCEC08C5A81FE3BF178EB7C799C394D85900DA1FFFA1E98A8C03008F25A35664A28D889D55B120726C9486053222642E60BE103143D6162B36D9303EFC96B9D5B6527D48B116B7AA33E6AB6F31955FB1AF7649DDF148DE9E3F0C17CD42DDEB0D678C2477183B9A617C274ABE1CF1571B2F7B007FF7C3210C4E412E870B23BC10A15889F9301D371E6DF61BCC7A4709E680D16E164687052A109EC3A2BA82353C454581BF92BF4F52D7670F80E561D66D0662B2D3F5D7259FCBC0403BE0D9CA417BA175E5B164D565BBC5F38DDB7E8DC6931463B17491E
      authenticator:
        etype: AES256_CTS_HMAC_SHA1_96 (18)
        kvno:
        cipher: 83F1AB5C4339E9BCAFA3281B73B933B01A17099F7CC720996BE9B3DCBC1E5237C6EC94BECCE3F1A7470CAF9E26230994722294055FB8690887DB501084543D6D25845D3C5D9C5F6CCD08E70B622DAA53A2E42A2FC3911C111142496CD396AE4B29D0D39495A9BC2D3DCAB11C178D0A7AE4ABEB27D3BA963413825EFF67BF1298B87D32070CDF23C58D5472AE3AA443F30E69A376363770904FFB293EB7A2A0087607ABA68254071E0FC7C77FD8369C5FC395A5934B539C038CB8AA774971CA4349BE00C97231BA193BEA30C1AFF62AB0F8F793ED92B81C7C24F517183A6078F96803027A95B45C773DF5DF014BC8EE91BB86718E9A5B8DF3A15D2D37156C3C4EAB866911FE70966D198B837D97FB970E1E94855E4E65F3901A5895A5F4F68825ADCAA514A30068DF14F4FBDE802412A4EF33F645022B631E97FB8E643DE04DC14B0E613C64D112763FD081A5762B362885D825EDCDBB3A0EEDFAED7D84C9BDE975F1AB95E2AD6AF278317BF27C45B7A64D6A673A1979004BC5E70E48F92CF72D77642B7A49D03C87E2994FD0DD2C3F1719704FBBBD98DACA9879CF291C4541FDA17643487F3AA6D1D6BAA083021E2BCF3525E7BE
    RawData: ...
RawData: ...
```

The key differences are the `ap-option` flags, the original has `128 - 0x80` set. There is most likely some other differences in the encrypted payloads but the general thing to take away here is the token format is the same, i.e. it's an `AP-REQ` inside `SPNEGO InitialContextToken` for the mech `1.2.840.113554.1.2.2`. I have no idea what that particular ap-options bit represents and it's not in any of the RFCs or parsers I've found around this field. I also don't think this is the problem but just some weird inconsistency.

The stack trace you got back is a failure of `pyspnego` to parse the incoming error message, this message looks like

```yaml
# python -m spnego --format yaml --token YHIGCS...
MessageType: SPNEGO InitialContextToken
Data:
  thisMech: Kerberos (1.2.840.113554.1.2.2)
  innerContextToken:
    MessageType: KRB-ERROR (30)
    Data:
      pvno: 5
      msg-type: KRB-ERROR (30)
      ctime:
      cusec:
      stime: '2022-01-10T22:42:09+00:00'
      susec: 51674
      error-code: KRB_ERR_GENERIC (60)
      crealm:
      cname:
      realm: DOMAIN.COM
      sname:
        name-type: NT-PRINCIPAL (1)
        name-string:
        - HTTP
        - krb-test.domain.com
      e-text:
      e-data:
    RawData: ...
RawData: ...
```

Unfortunately this doesn't really help as there is no error message for the generic error code but essentially this is not the type of token that `pyspnego` expects to receive back and thus rejects it. I'm unsure what the proper behaviour here is but regardless of whether it's a stacktrace or just properly formatted exception there was a failure on the authentication and the server is sending this generic error back rather than an empty 401 response which most typical servers do.

> The 401 error on the client side comes with an error on the server:
[Mon Jan 10 22:42:09.052037 2022] [auth_kerb:error] [pid 224941:tid 140553693796096] [client 203.0.113.42:57775] gss_accept_sec_context() failed: Unspecified GSS failure. Minor code may provide more information (, Invalid field length in token)

This is what looks to be the crux of the issue. It seems like the server is failing to parse the incoming token. This is surprising as the general structure of the tokens are the same and ultimately in both pyspnego's and pykerberos' examples they are both generated from a call to `GSSAPI` (`gss_init_sec_context`). This would lead me to think that the problem lies in the encrypted contents. When looking at the code for MIT krb5 that provides the GSSAPI library the only location this error appears is when [KG_BAD_LENGTH](https://github.com/krb5/krb5/blob/34625d594c339a077899fa01fc4b5c331a1647d0/src/lib/gssapi/krb5/gssapi_err_krb5.et#L31) is returned. This only only done in https://github.com/krb5/krb5/blob/34625d594c339a077899fa01fc4b5c331a1647d0/src/lib/gssapi/krb5/accept_sec_context.c#L475 and the indication is that it's a failure in the `authenticator` portion of the request which is unfortunately encrypted so we cannot see the differences in the plaintext token.

What it does mention is

> GSS_C_CHANNEL_BOUND_FLAG if matching channel bindings are present

Which would indicate this field contains the channel bindings structure that was added in release v0.12.0 (before pykerberos was changed to pyspnego). If you could only get a successful authentication attempt with v0.11.0 then this would indicate your server is unable to parse this value properly and omitting that step gets it working again. You can test this assumption by setting `HTTPKerberosAuth(send_cbt=False)` to disable the channel binding support.

Channel bindings is [part of the Kerberos RFC](https://datatracker.ietf.org/doc/html/rfc4121#section-4.1.1.2) so servers should support this but in the cases where it does not the `send_cbt` kwarg can be used to disable that option. You typically find CB required on Microsoft based endpoints where it is also known as Extended Protection for Authentication (EPA).

**Comment by twAlexM** [comment]
Thanks for the quick response — requesting `send_cbt=False` with 0.12.0 and also 0.14.0 produces successful authentication against our Apache mod_auth_kerb and Python flask_kerberos servers.

This option, and its default of True appears to be undocumented. Some notes about channel-binding support in the package documentation / changelog would be helpful to others tracking down this issue.

**Comment by jborean93** [comment]
https://github.com/requests/requests-kerberos/pull/174 adds more docs on this option and the change to send CBT in 0.12.0.

> Apache mod_auth_kerb 

I totally understand that you may not have control over this server but you should consider using [mod_auth_gssapi](https://github.com/gssapi/mod_auth_gssapi) which is the spiritual successor to `mod_auth_kerb`. The `mod_auth_kerb` plugin hasn't been updated since 2008 and using the GSSAPI one might even fix the problem with it not being able to process the CB data in the authenticator. It could also be that the web server is just running a really old version of MIT krb5 that doesn't support CB but in any case trying to update both is something I highly recommend. Even if you don't maintain the server I recommend letting who does know about this.

**Comment by twAlexM** [comment]
Thanks for the updated docs. I'll let the right people know about `mod_auth_gssapi`, as well. Cheers :)

**Comment by geofft** [comment]
I wonder if you ran into the thing I ran into with #180. It's not clear to me there are actually multiple hosts involved here, but I had a largely similar issue with AD FS (i.e., a Microsoft endpoint), and it was rejecting me if I provided _invalid_ CB data but not if I provided none at all.

On the other hand, if you've got a load-balancer or something else that terminates SSL set up, then channel binding simply won't work for you (but the server ought to be configurable to turn it off, then).
